### ISSUE TITLE
librarian-chef install [--no-git]

### ISSUE THREAD

**Issue by hedgehog** [applicationsonline/librarian#38]
remove .git folder after a cookbook has been downloaded.
This allows paranoid/fragile production deployments to track their cookbooks without getting tied in any git nested repos morass.  

Suited to production deployments where a cookbook change is highly unlikely, and where there is some need to ensure the cookbooks travel with the code base.

**Comment by yfeldblum** [comment]
Should this be the default? Should there be an option?

**Comment by hedgehog** [comment]
I don't have a strong opinion, though I will note that right now no `.git` is half the default, and half not :)  Given that 50:50 I'd prefer no `.git` be the default, that way it is consistent with the readme description of Librarian's intended use case - not you in-development-cookbooks.

Specifically, if you add a cookbook that has a dependency but you don't add the dependency, e.g `rsyslog`, then the dependency cookbook, `cron`, is added to `cookbooks` without a `.git` and the dependent cookbook, `rsyslog`, has a `.git`.

For others interested, an immediate workaround is:

```
cat <<'EOF' >.git/hooks/pre-commit
#!/usr/bin/env bash
set -e
for dir in `ls -d chef/cookbooks/**/.git`; do
    pushd $dir/..
    rm -rf .git
    popd
done
exit 0
EOF
chmod +x .git/hooks/pre-commit
```

But this you have to setup in each project and it is not tracked in your code base.

**Comment by yfeldblum** [comment]
@hedgehog,

I think in that particular example, `cron` is pulled from the default source (usually the Opscode Community Site API), while `rsyslog` is pulled from Git.

What's the morass with the Git nested repos? The general advice when using Librarian-Chef is specifically not to make any changes to any cookbooks in the Librarian-Chef-managed cookbooks directory (and there is an issue open to make that directory configurable). But what other scenarios are there that make working with Git-sourced cookbooks with Librarian-Chef problematic?

Also, what does Bundler do?

**Comment by hedgehog** [comment]
I'm quite happy to have the default Librarian use-case be a production deployment resolver and retriever of cookbooks, in fact I think that is the sweet spot, so I'm a big fan.  I tried to shoe-horn this into Bundler but... no-go.

A git repo within git repos can get pushed/pulled and checkedout in weird ways.  Apart from Gits assumptions, different tools chains can assume different things e.g. these are submodules.  Overall nested git repos are just not nice things to have unless you specifically have added them, and then manage them (ouch) as submodules.  One could leave the `.git` folders and suggest people organize their cookbooks in a separate folder outside of their project tree - but that seems to be a little too much of something... opinionated doesn't quite seem to be the word.
It would also break any production requirement that if the project's production repo is good, then the whole of the production stack is good.  Then you only need backups for one repo (I have some ideas for this aspect too and will open an issue for it)

Bundler+rbenv buy you a lot of flexibility in non-trivial projects about the Ruby runtime(s) you can use and the weight of the gemset that is loaded when certain components are loaded.  
Bundler+rbenv best practice in these cases is vendor the gem and bin files under the Gemfile, hence adding `vendor` and `bin` folders to the `.gitignore`.  In small.simple projects Bundler likely doesn't add much - but then in those projects not only wouldn't you need a Gemfile, you likely wouldn't need a Cheffile.  

So I imagine the default use case for a Cheffile is non-trivial production setting.

**Comment by yfeldblum** [comment]
Bundler does not remove `.git` or `.gitignore`.

Which tools assume that subdirs with `.git` are submodules, or otherwise break?

**Comment by hedgehog** [comment]
I've eliminated nested `.git` folders, so from memory: github and IDE's, and I think gitflow.  Anyway the main adv of removing  `.git` in what Librarian pulls down is that your cookbook is orders of magnitude lighter to move around.

A `--preserve-git` options won't hurt.  But, I'm not implementing this so that is easy to say.

**Comment by yfeldblum** [comment]
Whether or not this is in, it depends on #15, remembered configuration.

In other words, if you say you want to `librarian-chef install --strip-git`, that option should be stored & remembered the next time you run `librarian-chef`.

**Comment by till** [comment]
I think this should be default.

**Comment by yfeldblum** [comment]
@till Why do you think so?

**Comment by till** [comment]
I pretty much agree with what @hedgehog said. Less side-effects.

**Comment by yfeldblum** [comment]
Any concerns regarding cookbooks whose `metadata.rb` files shell out to `git`?

Don't know if there are any, but there are many ruby gems whose `.gemspec` files (`metadata.rb` equivalents) shell out to `git`.

**Comment by till** [comment]
Didn't know people did that – good Q. :)

**Comment by yfeldblum** [comment]
A lot of cookbooks out there do [this](https://github.com/opscode-cookbooks/apache2/blob/1.1.8/metadata.rb#L5) in their `metadata.rb` files:

```
long_description  IO.read(File.join(File.dirname(__FILE__), 'README.md'))
```

**Comment by hedgehog** [comment]
@yfeldblum, I've yet to see a `metadata.rb` shell out to `git`, this might be worth asking on the chef email list.  
Perhaps this can be placed on the list things Librarian assumes you don not do.  Which currently looks like:
- Use `git -ls` etc in your production cookbook recipes (if you do in development, use `librarian --preserve-git`)

**Comment by yfeldblum** [comment]
Closed by 3fb1f97171e38978541990f33ea80844ddacdf17.

For now, this is can be turned on, on a run-by-run basis, by setting the env var `LIBRARIAN_CHEF_INSTALL__STRIP_DOT_GIT=1`.

For now, this can be turned on, permanently for the project, by setting the config key `install.strip-dot-git` as follows:

```
librarian-chef config install.strip-dot-git 1 --local
```

For now, this can be turned off, permamently for the project, by deleting the config key `install.strip-dot-git` as follows:

```
librarian-chef config install.strip-dot-git --local --delete
```

For now, this can be checked as follows:

```
librarian-chef config install.strip-dot-git --local
```

Librarian and Librarian-Chef are not 1.0, so the names and the values may change at any time. Indeed, the values permitted for this key at the moment are anything at all (the `1` above is a placeholder), but that is probably going to change.

Note that this setting is _non-default_, since it does something extra.
